include/experimental/__p1673_bits/blas3_matrix_rank_k_update.hpp

```diff
@@ -0,0 +1,473 @@
+         class ElementType_C,
+         extents<>::size_type numRows_C,
+         extents<>::size_type numCols_C,
+         class Layout_C,
+         class Accessor_C,
+         class Triangle>
+void symmetric_matrix_rank_k_update(
+  ExecutionPolicy&& exec,
+  std::experimental::mdspan<ElementType_A, std::experimental::extents<numRows_A, numCols_A>, Layout_A, Accessor_A> A,
+  std::experimental::mdspan<ElementType_C, std::experimental::extents<numRows_C, numCols_C>, Layout_C, Accessor_C> C,
+  Triangle t)
+{
+
+  constexpr bool use_custom = is_custom_sym_mat_rank_k_update_avail<
+    decltype(execpolicy_mapper(exec)), void, decltype(A), decltype(C), Triangle
+    >::value;
+
+  if constexpr(use_custom) {
+    symmetric_matrix_rank_k_update(execpolicy_mapper(exec), A, C, t);
+  } else {
+    symmetric_matrix_rank_k_update(std::experimental::linalg::impl::inline_exec_t(), A, C, t);
+  }
+}
+
+template<class ElementType_A,
+         extents<>::size_type numRows_A,
+         extents<>::size_type numCols_A,
+         class Layout_A,
+         class Accessor_A,
+         class ElementType_C,
+         extents<>::size_type numRows_C,
+         extents<>::size_type numCols_C,
+         class Layout_C,
+         class Accessor_C,
+         class Triangle>
+void symmetric_matrix_rank_k_update(
+  std::experimental::mdspan<ElementType_A, std::experimental::extents<numRows_A, numCols_A>, Layout_A, Accessor_A> A,
+  std::experimental::mdspan<ElementType_C, std::experimental::extents<numRows_C, numCols_C>, Layout_C, Accessor_C> C,
+  Triangle t )
+{
+  symmetric_matrix_rank_k_update(std::experimental::linalg::impl::default_exec_t(), A, C, t);
+}
+
+
+// Rank-k update of a Hermitian matrix with scaling factor alpha
+
+template<class ScaleFactorType,
+	 class ElementType_A,
+         extents<>::size_type numRows_A,
+         extents<>::size_type numCols_A,
+         class Layout_A,
+         class Accessor_A,
+         class ElementType_C,
+         extents<>::size_type numRows_C,
+         extents<>::size_type numCols_C,
+         class Layout_C,
+         class Accessor_C,
+         class Triangle>
+void hermitian_matrix_rank_k_update(
+  std::experimental::linalg::impl::inline_exec_t&& /*exec*/,
+  ScaleFactorType alpha,
+  std::experimental::mdspan<ElementType_A, std::experimental::extents<numRows_A, numCols_A>, Layout_A, Accessor_A> A,
+  std::experimental::mdspan<ElementType_C, std::experimental::extents<numRows_C, numCols_C>, Layout_C, Accessor_C> C,
+  Triangle /* t */)
+{
+  using std::conj;
+  constexpr bool lower_tri = std::is_same_v<Triangle, lower_triangle_t>;
+  using size_type = typename extents<>::size_type;
+
+  for (size_type j = 0; j < C.extent(1); ++j) {
+    const size_type i_lower = lower_tri ? j : size_type(0);
+    const size_type i_upper = lower_tri ? C.extent(0) : j+1;
+    for (size_type i = i_lower; i < i_upper; ++i) {
+      for (size_type k = 0; k < A.extent(1); ++k) {
+        C(i,j) += alpha * A(i,k) * conj(A(j,k));
```

Comment:
Sorry I missed this in my earlier review!  `conj` always returns `std::complex`, even the overloads that take a non-`complex` argument.  This could cause build errors.

A straightforward fix for now would be to use a generic lambda.  We'll generally need to do this wherever we use `conj`, unless we know that the value type is `complex`.
```
constexpr bool A_is_complex =
  std::is_same_v<typename decltype(A)::value_type, std::complex<float>> ||
  std::is_same_v<typename decltype(A)::value_type, std::complex<double>> ||
  std::is_same_v<typename decltype(A)::value_type, std::complex<long double>>;
auto conj_if_needed = [](const auto A_jk) { 
  if constexpr (A_is_complex) {
    return conj(A_jk);
  } else {
    return A_jk;
  }
};
```

Reply:
ha! Thanks! 
Ok, so you mean this needs to be done *everywhere* we use `conj`? i will hunt those down in a separate PR maybe

Reply:
@mhoemmen btw, enjoy thanksgiving :)

Reply:
@fnrizzi wrote:
> For instance `rank_2k` uses that, in fact I used that as starting point :)

No promise that existing code was correct ; - )

There are two issues with using `conj` "generically" (for any element type):

1. It always returns `std::complex<U>` for some `U`.  The resulting arithmetic expression may not compile.
2. It only has overloads for floating-point and integer types.  Thus, it won't compile for types that are not implicitly convertible to one of these.